Assets/CodeBase/Entities/player/Player.cs

```diff
@@ -78,27 +94,70 @@ void Start()
         inputProfile.addListener(InputEvent.Key, PlayerInputProfile.moveDown, moveDown);
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.moveUp, stopClimbing);
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.moveDown, stopClimbing);
-        inputProfile.addListener(InputEvent.Key, PlayerInputProfile.jump, jump);
+        inputProfile.addListener(InputEvent.Down, PlayerInputProfile.jump, jump);
 
         //Listeners for abilities.
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.toggleIce, toggleIce);
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.toggleFire, toggleFire);
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.toggleWind, toggleWind);
         inputProfile.addListener(InputEvent.Up, PlayerInputProfile.toggleEarth, toggleEarth);
 
+        //shift key
+        inputProfile.addListener(InputEvent.Down, PlayerInputProfile.shift, shift);
+
         currentAbility = ActiveAbility.NORMAL;
         this.enabled = false;
         Controller.instance.stateMachine.AddStateListener(onStateChange);
 
     }
 
     // Update is called once per frame
-    void Update()
+    void FixedUpdate()
     {
         inputProfile.checkInput();
+        //print("currentAbility: " + currentAbility);
 
-        if (_isHuggingWall && !_isClimbing && currentAbility.Equals(ActiveAbility.EARTH))
-        {
+        //print("jumpSpeed: " + jumpSpeed);
+
+        //code for elemental power logic goes here
+        //ice...
+        if (currentAbility.Equals(ActiveAbility.ICE)) {
+            //slow player down and lock down jump ability
+            moveSpeed = initMoveSpeed / 2f; //halve(?) movement speed
+            jumpSpeed = 0; //means: can't jump
+        }
+        else {
+            moveSpeed = initMoveSpeed;
+            jumpSpeed = initJumpSpeed;
+        }
+
+        //fire...
+        if (currentAbility.Equals(ActiveAbility.FIRE)) {
+            if (_canDash) {
+                float moveH = Input.GetAxisRaw("Horizontal");
+                float moveV = Input.GetAxisRaw("Vertical");
+
+                //print(moveH + ", " + moveV);
+
+                //dash effect from fire, restrict to only in the air for simpler implementation
+                if (_shiftPressed && !_isGrounded) {
+                    _dashing = true;
+                    _canDash = false;
+                    print("add fire dash force!");
+
+                    //add extra upwards force to push against gravity
+                    playerRigidBody.velocity = new Vector2(moveH * fireDashSpeed, moveV * fireDashSpeed + 3f);
```

Comment:
Setting the gravityScale to 0 for the playerRigidBody would be a decent way to prevent gravity from affecting dashes (setting it back to 1 once the dash is complete. This method worked well for climbing with the earth ability).